examples/wave/wave-opt-3d.py

```diff
@@ -0,0 +1,821 @@
+                                          existing_ok=True)
+
+            for i, var_names in enumerate((priv_cse_kernel2_vars,
+                                           priv_cse_kernel3_vars)):
+                for var_name in var_names:
+                    knl = lp.assignment_to_subst(knl, var_name)
+                for var_name in var_names[::-1]:
+                    knl = precompute_for_single_kernel(
+                        knl, t_unit.callables_table, f"{var_name}_subst",
+                        sweep_inames=(),
+                        temporary_address_space=lp.AddressSpace.PRIVATE,
+                        compute_insn_id=f"cse_knl{i+2}_prcmpt_{var_name}")
+
+            knl = lp.map_instructions(knl,
+                      " or ".join(f"writes:{var_name}"
+                                  for var_name in cse_kernel1),
+                      lambda x: x.tagged(lp.LegacyStringInstructionTag("cse_knl1")))
+
+            knl = lp.map_instructions(knl,
+                      " or ".join([f"writes:{var_name}"
+                                   for var_name in cse_kernel2]
+                                  + ["id:cse_knl2_prcmpt_*"]),
+                      lambda x: x.tagged(lp.LegacyStringInstructionTag("cse_knl2")))
+
+            knl = lp.map_instructions(knl,
+                      " or ".join([f"writes:{var_name}"
+                                   for var_name in cse_kernel3]
+                                  + ["id:cse_knl3_prcmpt_*"]),
+                      lambda x: x.tagged(lp.LegacyStringInstructionTag("cse_knl3")))
+
+            # }}}
+
+            # {{{ Stats
+
+            if 1:
+                from loopy.kernel.array import ArrayBase
+                from pytools import product
+                t_unit = t_unit.with_kernel(knl)
+
+                op_map = lp.get_op_map(t_unit,
+                                       subgroup_size=32)
+                f64_ops = op_map.filter_by(dtype=[np.float64],
+                                           kernel_name="_pt_kernel").eval_and_sum({})
+
+                # {{{ footprint gathering
+
+                nfootprint_bytes = 0
+
+                for ary in knl.args:
+                    if (isinstance(ary, ArrayBase)
+                            and ary.address_space == lp.AddressSpace.GLOBAL):
+                        nfootprint_bytes += (product(ary.shape)
+                                             * ary.dtype.itemsize)
+
+                for ary in knl.temporary_variables.values():
+                    if ary.address_space == lp.AddressSpace.GLOBAL:
+                        # global temps would be written once and read once
+                        nfootprint_bytes += (2 * product(ary.shape)
+                                             * ary.dtype.itemsize)
+
+                # }}}
+
+                print("Double-prec. GFlOps:", f64_ops * 1e-9)
+                print("Footprint GBs:",  nfootprint_bytes * 1e-9)
```

Comment:
@inducer: Could you take a look at the way we are gathering these metrics. Any issues?